### ISSUE TITLE
Arduino Due first.gpr demo is outdated and doesn't build

### ISSUE THREAD

**Issue by Electromechanical-Supergiant** [simonjwright/cortex-gnat-rts#23]
I couldn't get it to build. Upon closer inspection of first_task.adb, it is `with`ing package `Registers`, which doesn't appear to be defined anywhere that I could find.

I was trying to build it to get an idea of how to write to registers, which is actually my real issue here. The demo uses code such as `PIOB.CODR := (Output_Pin => 1, others => 0);` This may work with the current atsam3x8e* files, but is a bit clumsy if you need to write more than one pin at a time. Ideally I'd like to be able to use a bit mask to write to a register.

I might edit the BSP files by hand to make the register definitions simpler, as well as moving definitions of their addresses to a linker script (where I believe they should belong). If you're interested in what changes I make, I can fork and make a PR later.

**Comment by simonjwright** [comment]
Sorry about that. A while ago I abandoned my own arduino-due-bsp, on the grounds that the definitions would be extremely difficult and laborious to maintain by hand, when an automated approach is available via [SVD2Ada](https://github.com/AdaCore/svd2ada). I do need to add somewhere the fact I’ve done that, and the parameters I used to generate the `atsam3x8e/` files.

I’ve taken the opportunity to update the demos in this directory to the current spec.

>I was trying to build it to get an idea of how to write to registers, which is actually my real issue here. The demo uses code such as `PIOB.CODR := (Output_Pin => 1, others => 0);` This may work with the current atsam3x8e* files, but is a bit clumsy if you need to write more than one pin at a time. Ideally I'd like to be able to use a bit mask to write to a register.

This now reads
```ada
PIOB_Periph.CODR.Arr := (Output_Pin => 1, others => 0);
```
and, if you wanted to set more than one bit, it’d be, for bits 1, 2, 8 and 29,
```ada
    PIOB_Periph.CODR.Arr := (1 | 2 | 8 | 29 => 1, others => 0);
```

>I might edit the BSP files by hand to make the register definitions simpler, as well as moving definitions of their addresses to a linker script (where I believe they should belong). If you're interested in what changes I make, I can fork and make a PR later.

You may have gathered from the above that that’s not a change I’d wish to make.

I don’t see why the address definitions would be better in the linker script? This is certainly the approach adopted in the [Ada and SPARK on ARM Cortex-M book Chapter 5](http://www.inspirel.com/articles/Ada_On_Cortex_Digital_Output.html):

>We will use a third approach, which allows separation of concerns: the source code will be responsible for business logic and the linker script will take responsibility for memory layout and detailed object placement. The advantage of this approach is that handling of addresses and memory locations is kept in a single place and the program source code is not polluted with hardware details.

but it seems to me that you then have to do the work twice: you'd need to specify a symbol `piob_codr` with value `0x400E1034` in the linker script, and then import this in your Ada source:
```ada
PIOB_CODR : Interfaces.Unsigned_32
   with Import,
   Convention => C,
   External_Name => "piob_codr";
```

**Comment by Electromechanical-Supergiant** [comment]
Thanks, the example compiles now. Actually, I had to edit line 27 in the `Builder` package of first.gpr as I got the following error:

```
first.gpr:27:11: Default_Switches forbidden in presence of Global_Compilation_Switches. Use Switches instead.
gprbuild: *** illegal combination of Builder attributes
``` 

As you guessed, I found the style of putting hardware addresses in the linker script from that book. I am attracted to the separation of concerns it provides, but you're right in that there is some level of duplication involved in doing so.

I recently completed that book and it's been my first foray into the embedded world. I'm also still relatively new with Ada, so I appreciate the detail you've given in your response and I'll defer to your judgement.

I initially tried to build my own runtime for the Due without FreeRTOS, but after spending a day on it without success I decided to use your runtime here. The underlying RTOS will come in handy later for me anyway. Thanks for all your work here (and also all your answers and comments on Stack Overflow, you've already helped me more than you know ;) ).

**Comment by simonjwright** [comment]
Thanks for the reply & the note. I was using the latest gprbuild from Github, which does *not* complain about `Default_Switches`! Also I find the whole business of what is & what isn’t allowed in `package Builder` inscrutable: you’re allowed to use `Switches (others)` but not `Global_Compilation_Switches (others)`, has to be `("ada")`.

**Comment by simonjwright** [comment]
Fixed in a5ba4ad33aad93ce46e627414bd144eedd50b734, e58cef96f0095c6fbcce4f0bd6d2d617c381b0d0, 71edf97ac94ac521b6299cde575edc9e09551587.
